File: toneguard-mcp/merge.py
"""ToneGuard merge strategies — pure functions for conflict resolution.

Each function takes (local, remote) and returns the merged result.
Port of src/sync/merge.js — must produce identical output for the same inputs.
"""

from __future__ import annotations


def merge_decisions(local: list | None, remote: list | None) -> list:
    """Merge decision arrays: union by (timestamp+action), sort newest-first, trim to 100."""
    all_items = list(local or []) + list(remote or [])
    seen: set[str] = set()
    deduped: list[dict] = []

    for d in all_items:
        key = (d.get("timestamp") or "") + "|" + (d.get("action") or "")
        if key in seen:
            continue
        seen.add(key)
        deduped.append(d)

    deduped.sort(key=lambda d: d.get("timestamp") or "", reverse=True)
    return deduped[:100]


def merge_voice_samples(local: list | None, remote: list | None) -> list:
    """Merge voice sample arrays: deduplicate by text content, keep newest 30."""
    all_items = list(local or []) + list(remote or [])
    seen: set[str] = set()
    deduped: list[dict] = []

    for s in all_items:
        key = s.get("text") or ""
        if key in seen:
            continue
        seen.add(key)
        deduped.append(s)

    deduped.sort(key=lambda s: s.get("timestamp") or "", reverse=True)
    return deduped[:30]


def merge_relationships(local: dict | None, remote: dict | None) -> dict:
    """Merge relationship maps: per-key, take max messageCount and latest lastSeen."""
    local_map = local or {}
    remote_map = remote or {}
    merged: dict = {}
    all_keys = set(list(local_map.keys()) + list(remote_map.keys()))

    for key in all_keys:
        l = local_map.get(key, {"messageCount": 0, "lastSeen": None})
        r = remote_map.get(key, {"messageCount": 0, "lastSeen": None})

        merged[key] = {
            "messageCount": max(l.get("messageCount") or 0, r.get("messageCount") or 0),
            "lastSeen": l.get("lastSeen") if (l.get("lastSeen") or "") > (r.get("lastSeen") or "") else r.get("lastSeen"),
        }

    return merged


def merge_custom_rules(local: dict | None, remote: dict | None) -> dict:
    """Merge custom rules: last-write-wins based on updatedAt timestamp."""
    local_val = local or {"rules": "", "updatedAt": ""}
    remote_val = remote or {"rules": "", "updatedAt": ""}

    if (remote_val.get("updatedAt") or "") > (local_val.get("updatedAt") or ""):
        return {
            "rules": remote_val.get("rules", ""),
            "source": "remote",
            "updatedAt": remote_val.get("updatedAt", ""),
        }
    return {
        "rules": local_val.get("rules", ""),
        "source": "local",
        "updatedAt": local_val.get("updatedAt", ""),
    }


def _merge_by_mode(a: dict | None, b: dict | None) -> dict:
    """Merge byMode dicts: per-key max."""
    mode_a = a or {}
    mode_b = b or {}
    merged = {**mode_a}

    for key, val in mode_b.items():
        merged[key] = max(merged.get(key) or 0, val or 0)

    return merged


def merge_stats_history(local: list | None, remote: list | None) -> list:
    """Merge stats history: union by weekStart, take higher counts per week, trim to 12."""
    local_arr = local or []
    remote_arr = remote or []
    by_week: dict[str, dict] = {}

    for week in [*local_arr, *remote_arr]:
        key = week.get("weekStart") or ""
        existing = by_week.get(key)

        if existing is None:
            by_week[key] = {**week}
        else:
            by_week[key] = {
                "weekStart": key,
                "checked": max(existing.get("checked") or 0, week.get("checked") or 0),
                "flagged": max(existing.get("flagged") or 0, week.get("flagged") or 0),
                "accepted": max(existing.get("accepted") or 0, week.get("accepted") or 0),
                "edited": max(existing.get("edited") or 0, week.get("edited") or 0),
                "dismissed": max(existing.get("dismissed") or 0, week.get("dismissed") or 0),
                "byMode": _merge_by_mode(existing.get("byMode"), week.get("byMode")),
            }

    merged = list(by_week.values())
    merged.sort(key=lambda w: w.get("weekStart") or "")
    return merged[-12:]

File: src/sync/merge.js
// ToneGuard merge strategies — pure functions for conflict resolution.
// Each function takes (local, remote) and returns the merged result.

/**
 * Merge decision arrays: union by (timestamp + action), sort newest-first, trim to 100.
 */
function mergeDecisions(local, remote) {
  const all = [...(local || []), ...(remote || [])];
  const seen = new Set();
  const deduped = [];

  for (const d of all) {
    const key = (d.timestamp || "") + "|" + (d.action || "");
    if (seen.has(key)) continue;
    seen.add(key);
    deduped.push(d);
  }

  deduped.sort((a, b) => {
    const ta = a.timestamp || "";
    const tb = b.timestamp || "";
    return tb.localeCompare(ta);
  });

  return deduped.slice(0, 100);
}

/**
 * Merge voice sample arrays: deduplicate by text content, keep newest 30.
 */
function mergeVoiceSamples(local, remote) {
  const all = [...(local || []), ...(remote || [])];
  const seen = new Set();
  const deduped = [];

  for (const s of all) {
    const key = s.text || "";
    if (seen.has(key)) continue;
    seen.add(key);
    deduped.push(s);
  }

  deduped.sort((a, b) => {
    const ta = a.timestamp || "";
    const tb = b.timestamp || "";
    return tb.localeCompare(ta);
  });

  return deduped.slice(0, 30);
}

/**
 * Merge relationship maps: per-key, take max messageCount and latest lastSeen.
 */
function mergeRelationships(local, remote) {
  const localMap = local || {};
  const remoteMap = remote || {};
  const merged = {};
  const allKeys = new Set([...Object.keys(localMap), ...Object.keys(remoteMap)]);

  for (const key of allKeys) {
    const l = localMap[key] || { messageCount: 0, lastSeen: null };
    const r = remoteMap[key] || { messageCount: 0, lastSeen: null };

    merged[key] = {
      messageCount: Math.max(l.messageCount || 0, r.messageCount || 0),
      lastSeen: (l.lastSeen || "") > (r.lastSeen || "") ? l.lastSeen : r.lastSeen
    };
  }

  return merged;
}

/**
 * Merge custom rules: last-write-wins based on updatedAt timestamp.
 * Returns { rules, source } where source is "local" or "remote".
 */
function mergeCustomRules(local, remote) {
  const localVal = local || { rules: "", updatedAt: "" };
  const remoteVal = remote || { rules: "", updatedAt: "" };

  if ((remoteVal.updatedAt || "") > (localVal.updatedAt || "")) {
    return { rules: remoteVal.rules, source: "remote", updatedAt: remoteVal.updatedAt };
  }
  return { rules: localVal.rules, source: "local", updatedAt: localVal.updatedAt };
}

/**
 * Merge stats history: union by weekStart, take higher counts per week.
 */
function mergeStatsHistory(local, remote) {
  const localArr = local || [];
  const remoteArr = remote || [];
  const byWeek = new Map();

  for (const week of [...localArr, ...remoteArr]) {
    const key = week.weekStart || "";
    const existing = byWeek.get(key);

    if (!existing) {
      byWeek.set(key, { ...week });
    } else {
      byWeek.set(key, {
        weekStart: key,
        checked: Math.max(existing.checked || 0, week.checked || 0),
        flagged: Math.max(existing.flagged || 0, week.flagged || 0),
        accepted: Math.max(existing.accepted || 0, week.accepted || 0),
        edited: Math.max(existing.edited || 0, week.edited || 0),
        dismissed: Math.max(existing.dismissed || 0, week.dismissed || 0),
        byMode: mergeByMode(existing.byMode, week.byMode)
      });
    }
  }

  const merged = [...byWeek.values()];
  merged.sort((a, b) => (a.weekStart || "").localeCompare(b.weekStart || ""));

  return merged.slice(-12);
}

function mergeByMode(a, b) {
  const modeA = a || {};
  const modeB = b || {};
  const merged = { ...modeA };

  for (const [key, val] of Object.entries(modeB)) {
    merged[key] = Math.max(merged[key] || 0, val || 0);
  }

  return merged;
}

// Last-write-wins on updatedAt. Unlike raw voice_samples (CRDT-union),
// the fingerprint is a derived artifact regenerated on demand, so newest
// wins is the right semantics. Returns null when both sides are null.
function mergeVoiceFingerprint(local, remote) {
  if (local == null && remote == null) return null;
  if (local == null) return remote;
  if (remote == null) return local;
  const localAt = local.updatedAt || "";
  const remoteAt = remote.updatedAt || "";
  return remoteAt > localAt ? remote : local;
}

if (typeof globalThis !== "undefined") {
  globalThis.__toneGuardMerge = {
    mergeDecisions,
    mergeVoiceSamples,
    mergeVoiceFingerprint,
    mergeRelationships,
    mergeCustomRules,
    mergeStatsHistory
  };
}
